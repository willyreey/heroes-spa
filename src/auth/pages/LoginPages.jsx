import { useState } from "react"
import { useContext } from "react"
import { useNavigate } from "react-router-dom"
import { AuthContext } from "../context/AuthContext"

export const LoginPages = () => {

  const {login} = useContext(AuthContext)

  const navigate = useNavigate()

  const [loginName, setLoginName] = useState('')

  const handleInputChange = ({target}) => {
    setLoginName(target.value)
  }

  const onLogin = (e) =>{
    e.preventDefault()

    const lastPath = localStorage.getItem('lastPath') || '/' 
    
    login(loginName)
    navigate(lastPath,{
      replace: true
    })
  }

  return (
    <div className="container mt-5 main">
      
      <h1 className="sign" align="center">App Super Hero</h1>
      <hr />
      <div className="row">
        <form onSubmit={onLogin} className="form1">
            <input className="form-control mx-auto un" type="text" placeholder="Ingrese su Nombre" value={loginName} onChange={handleInputChange} required/>
            <button 
              className="submit mt-4"
            >
              Entrar
            </button>
        </form>
      </div>
    </div>
  )
}
