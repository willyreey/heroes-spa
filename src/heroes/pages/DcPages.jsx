import { HeroList } from "../components"

export const DcPages = () => {
  return (
    <>
      <hr />
      <h1>DcPages</h1>
      <hr />
      <HeroList publisher={'DC Comics'}/>
    </>
  )
}
