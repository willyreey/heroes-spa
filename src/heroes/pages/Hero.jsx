import { useMemo } from "react"
import { Navigate, useNavigate, useParams } from "react-router-dom"
import { getHeroById } from "../helpers/getHeroById"
import '../components/HeroCard.css'


export const Hero = () => {

  const { id } = useParams()
  const hero = useMemo(() =>getHeroById(id),[id])

  const navigate = useNavigate()

  const onNavigateBack = () =>{
    navigate(-1);
  }

  if( !hero ){
    return <Navigate to="/marvel"/>
  }
  console.log(hero);

  return (
    <div className="justify-content-center animate__animated animate__slideInLeft">
        <div className="row mx-auto mt-5">
          <div className="col-4">
          <img className="img-thumbnail" src={`../../assets/heroes/${id}.jpg`} alt="Card image cap" />
          </div>
         
          <div className="col-4">
            <h2>{hero.superhero}</h2>
            <small>{hero.characters}</small>
            <hr />
            <h5>{hero.publisher}</h5>
            <p>Primera apariciÃ³n: <br />{hero.first_appearance}</p>
            {
              (hero.alter_ego !== hero.characters)
              && <p>{hero.characters}</p>
            }
            <hr />
            <button className="btn btn-success" onClick={onNavigateBack}>Atras</button>
          </div>
          
        </div>
    </div>
  )
}
