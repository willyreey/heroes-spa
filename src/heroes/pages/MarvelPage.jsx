import { HeroList } from "../components"

export const MarvelPage = () => {
  return (
    <>
      <hr />
      <h1>MarvelPage</h1>
      <hr />
      <HeroList publisher={'Marvel Comics'}/>
    </>
  )
}
