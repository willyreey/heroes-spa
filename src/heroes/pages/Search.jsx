import { useLocation, useNavigate } from "react-router-dom"
import queryString from 'query-string'
import { useForm } from "../../hooks/useForm"
import { HeroCard } from "../components"
import { getHeroByName } from "../helpers"

export const Search = () => {

  const navigate = useNavigate()
  const location = useLocation()
  const {q = ''} = queryString.parse(location.search)
  const heroes = getHeroByName(q)


  const onSearchSubmit = (e) => {
    e.preventDefault()
    navigate(`?q=${searchText}`)
  }

  const onNavigateBack = () =>{
    navigate(-1);
  }

  const { searchText, onInputChange} = useForm({
    searchText: q
  })
  return (
    <>
      <hr />
      <div className="row">
        <div className="col-5">
        <h4>Búsqueda</h4>
        <hr />
          <form onSubmit={onSearchSubmit} aria-label="form">
            <input 
              type="text" 
              placeholder="Buscar Héroe" 
              className="form-control" 
              name="searchText" 
              autoComplete="off" 
              value={searchText} 
              onChange={onInputChange}
            />
            <button className="btn btn-primary mt-3">
              Buscar
            </button>
            <button className="btn btn-success mt-3 mx-2" onClick={onNavigateBack}>
              Atras
            </button>
          </form>

        </div>
        
        <div className="col-7">
          <h4>Resultados</h4>
          <hr />
          <div 
            className="alert alert-primary animate__animated animate__fadeIn" 
            style={{ display: q !== '' ? 'none' : ''}}
          >
            Buscar un Héroe
          </div>

          <div 
            className="alert alert-danger animate__animated animate__fadeIn" 
            style={{ display: (heroes.length === 0) && (q.length > 0) ? '' : 'none'}}
            aria-label="alert-danger"
          >
            No se encontro un héroe con el nombre <b>{q}</b>.
          </div>
          <div className="row animate__animated animate__fadeIn">
            {
              heroes.map(hero => (
                <HeroCard key={hero.id} {...hero} />
              ))
            }
          </div>
        </div>
      </div>
    
    </>
  )
}
