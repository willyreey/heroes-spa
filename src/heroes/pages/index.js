export * from './DcPages'
export * from './MarvelPage'
export * from './Search'
export * from './Hero'


