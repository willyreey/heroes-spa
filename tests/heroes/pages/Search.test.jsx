import { fireEvent, render, screen } from "@testing-library/react"
import { MemoryRouter } from "react-router-dom"
import { Search } from "../../../src/heroes/pages/Search"

const mockedSetSearchParams = jest.fn()

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockedSetSearchParams
}))

describe('Pruebas en Search', () => { 

    beforeEach(() => jest.clearAllMocks)

    test('Debe de mostrarse correctamente con valores por defecto', () => { 

        const {container} = render(
            <MemoryRouter>
                <Search/>
            </MemoryRouter>
        )
        expect(container).toMatchSnapshot()
    
     })

    test('Debe de mostrar a batman y el input con el valor del queryString', () => { 

        render(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Search/>
            </MemoryRouter>
        )

        const input = screen.getByRole('textbox')
        expect(input.value).toBe('batman')

        const img = screen.getByRole('img')
        expect(img.src).toContain('assets/heroes/dc-batman.jpg')

        const alertDanger = screen.getByLabelText('alert-danger')
        expect(alertDanger.style.display).toBe('none') 

    })

    test('Debe de mostrar un error si no se encuentra el hero', () => { 

        render(
            <MemoryRouter initialEntries={['/search?q=gatubela']}>
                <Search/>
            </MemoryRouter>
        )

        const alertDanger = screen.getByLabelText('alert-danger')
        expect(alertDanger.style.display).toBe('') 
    
    })

    test('Debe de llamar el navigate a la pantalla nuevas', () => { 

        render(
            <MemoryRouter initialEntries={['/search']}>
                <Search/>
            </MemoryRouter>
        )

        const input = screen.getByRole('textbox')
        fireEvent.change(input, {target:{name: 'searchText', value: 'superman'}})
        const form = screen.getByLabelText('form')
        fireEvent.submit( form )
        expect( mockedSetSearchParams ).toHaveBeenCalled(); 
        expect( mockedSetSearchParams ).toHaveBeenCalledWith('?q=superman')
        
    
    })

 })