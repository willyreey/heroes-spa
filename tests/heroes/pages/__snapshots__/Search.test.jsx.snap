// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en Search Debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Búsqueda
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Buscar Héroe"
          type="text"
          value=""
        />
        <button
          class="btn btn-primary mt-3"
        >
          Buscar
        </button>
        <button
          class="btn btn-success mt-3 mx-2"
        >
          Atras
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        class="alert alert-primary animate__animated animate__fadeIn"
      >
        Buscar un Héroe
      </div>
      <div
        aria-label="alert-danger"
        class="alert alert-danger animate__animated animate__fadeIn"
        style="display: none;"
      >
        No se encontro un héroe con el nombre 
        <b />
        .
      </div>
      <div
        class="row animate__animated animate__fadeIn"
      />
    </div>
  </div>
</div>
`;
